W-13743400-Add-sizing-guide-in-1-5-CP (#407)
mulesoft/docs-gateway: gateway/1.5/modules/ROOT/pages/flex-sizing-guide.adoc

= Reviewing Resource Sizing 
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:table-caption!:

> Before getting started with Flex Gateway, you must analyze and choose your configuration to obtain the best performance and auto-scaling rate.

Reviewer: ```suggestion
Before getting started with Flex Gateway, analyze and choose your configuration to obtain the best performance and auto-scaling rate.
```

Author: OK! Thanks!

> When assigning resources for a node, you must consider:

Reviewer: ```suggestion
When assigning resources for a node, consider:
```

Author: OK! Thanks!

* Number of APIs
* Amount of memory the node needs
* Processing power for the node
* Number of policies in each API
* Amount of traffic each node handles
* Amount of time the node takes to answer the request (latency)
* Type of traffic the node handles, and within the traffic:
** The number of requests per second the node handles
** Size of each request (bytes)

> == Choose Appropriate Resources
> The following table guides you to know the resources you must allocate to your node according to your environment:

Reviewer: ```suggestion
== Choose Appropriate Resources

The following table guides you to know the resources you must allocate to your node according to your environment:
```
I think you need this empty line for the Header to render properly. Please confirm in a local build.

Author: OK! Thanks!

[options="header"]
|====
| Node Size | CPU | RAM 
| Small | 1-2 cores | 2-4 GB
| Medium | 2-4 cores | 4-8 GB
| Large | 8-16 cores | 16-32 GB	
|====

NOTE: Nodes belonging to a cluster must be of the same size.

== Choose Your Cluster Size

> You can choose the appropriate cluster size with the following guiding table:

Reviewer: ```suggestion
The following guiding table helps you choose the appropriate cluster size:
```

Author: OK! Thanks!

[options="header"]
|====
| Node Size | Number of APIs | Number of Policies | Expected Latency | Expected Throughput | Target Environment
| Small | < 100 | 4 | < 100 ms | < 500 RPS | Dev/test cluster
| Medium | < 1000 | 4 | < 20 ms | < 2500 RPS | Production cluster
| Large | < 10000 | 4 | < 10 ms | < 10000 RPS | Mission-critical clusters
|====

NOTE: A high CPU can lead to increased latency. Proceed with horizontal scaling or increase the node size. High memory consumption can lead to out-of-memory errors. Increase the memory allocation or the node size. Horizontal scaling can fail helping in high-memory situations.

> For the guiding table, the average request size is 1kB and 4 policies are implemented. Adjust your parameters depending on your use case. 

Reviewer: ```suggestion
The guiding table considers a cluster with an average request size of 1kB and 4 policies implemented. Adjust your parameters depending on your use case. 
```

Author: Great wording! Thanks!

> The 4 out-of-the-box policies commonly used in these guidelines are:

Reviewer: ```suggestion
The most commonly used out-of-the-box policies, which are considered in these guidelines, are:
```

Author: OK! Thanks!

* Message logging
* Rate limiting
* JWT validation
* Header injection

> NOTE: Using TLS highly affects the response time and the process consumption.

Reviewer: ```suggestion
Using TLS highly affects the response time and the process consumption.
```

Author: OK! Thanks!

> 

Reviewer: This last section seems a bit out of order, I'd suggest the following order: 

== Choose Your Cluster Size

High Cpu... 
High Memory... 
The following table... 
[table] 
This table considers... 
The most commonly used... 
...

Author: It's interesting the order you proposed because you place the information about the CPU, memory, scaling, etc. before the table, and I put it after the table.
From my point of view, the note **AFTER** the table adds data for choosing a different node size in case the reader follows the table and the results do not fulfill their expectations.
The note **BEFORE** the table provides an explanation to understand the variables in the table beyond our recommendations. 
Said that what option do you think is best? Am I right in my assumptions?
Thanks!

* Number of APIs
* Amount of memory the node needs
* Processing power for the node
* Number of policies in each API
* Amount of traffic each node handles
* Amount of time the node takes to answer the request (latency)
* Type of traffic the node handles, and within the traffic:
** The number of requests per second the node handles
** Size of each request (bytes)

== Choose Appropriate Resources

The following table guides you to know the resources you must allocate to your node according to your environment:

[options="header"]
|====
| Node Size | CPU | RAM 
| Small | 1-2 cores | 2-4 GB
| Medium | 2-4 cores | 4-8 GB
| Large | 8-16 cores | 16-32 GB	
|====

NOTE: Nodes belonging to a cluster must be of the same size.

== Choose Your Cluster Size

The following guiding table helps you choose the appropriate cluster size:

[options="header"]
|====
| Node Size | Number of APIs | Number of Policies | Expected Latency | Expected Throughput | Target Environment
| Small | < 100 | 4 | < 100 ms | < 500 RPS | Dev/test cluster
| Medium | < 1000 | 4 | < 20 ms | < 2500 RPS | Production cluster
| Large | < 10000 | 4 | < 10 ms | < 10000 RPS | Mission-critical clusters
|====

> NOTE: A high CPU can lead to increased latency. Proceed with horizontal scaling or increase the node size. High memory consumption can lead to out-of-memory errors. Increase the memory allocation or the node size. Horizontal scaling can fail helping in high-memory situations.

Reviewer: ```suggestion
High CPU usage can lead to increased latency. In such a case, proceed with horizontal scaling or increase the node size. 

High memory consumption can lead to out-of-memory errors. To solve this issue, increase the memory allocation or the node size. Horizontal scaling might not help solve issues in this situation.
```

Author: OK! Thanks!

= Reviewing Resource Sizing 
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:table-caption!:

Before getting started with Flex Gateway, analyze and choose your configuration to obtain the best performance and auto-scaling rate.

When assigning resources for a replica node, consider:

* Number of APIs
* Amount of memory the node needs
* Processing power for the node
* Number of policies in each API
* Amount of traffic each node handles
* Amount of time the node takes to answer the request (latency)
* Type of traffic the node handles, and within the traffic:
** The number of requests per second the node handles
** Size of each request (bytes)

== Choose Appropriate Resources

The following table guides you to know the resources you must allocate to your replica node according to your environment:

[options="header"]
|====
| Node Size | CPU | RAM 
| Small | 1-2 cores | 2-4 GB
| Medium | 2-4 cores | 4-8 GB
| Large | 8-16 cores | 16-32 GB	
|====

NOTE: Nodes belonging to a cluster must be of the same size.

== Choose Your Cluster Size

> The following guiding table helps you choose the appropriate cluster size:

Reviewer 2: "...appropriate node and cluster size..."

Author: OK! Thanks!

[options="header"]
|====
| Node Size | Number of APIs | Number of Policies | Expected Latency | Expected Throughput | Target Environment
| Small | < 100 | 4 | < 100 ms | < 500 RPS | Dev/test cluster
| Medium | < 500 | 4 | < 20 ms | < 2500 RPS | Production cluster
| Large | > 500 | 4 | < 10 ms | < 10000 RPS | Mission-critical clusters

> |====

Reviewer 2: The figures above are per node. API configurations will be replicated to all nodes in the cluster. So, for example, having 10 APIs deployed with 3 policies each will mean that all 10 APIs and their policies will be deployed to each of the nodes in the cluster.

Traffic will be distributed across all nodes in the cluster. Thus, a Small cluster node supports up to 500 RPS, but if the cluster has two nodes then the total traffic that the cluster allows is 1000 RPS (500 RPS each node).  Exceeding the recommended traffic per node will eventually cause increased CPU usage and the recommendation is to either increase the node size (for all nodes in the cluster) or scale horizontally.

Author: OK! Thanks for the clarification!
